File: lab2/rework/cpp_solution/Highlighter.cpp
#include "Highlighter.hpp"

Highlighter::Highlighter(Document* chs)
{
    chdr = chs;
    lex = new Lexer(chs);
}


Highlighter::Highlighter(Document* chs, Lexer* l)
{
    chdr = chs;
    lex = l;
}


void Highlighter::addKeywords(std::vector<std::string>&& kws)
{
    for (auto i = kws.begin(); i != kws.end(); ++i)
        lex->addKeyword(*i);
}

Document Highlighter::highlightCharStream()
{
    Token::Token currTkn = lex->scan();
    while (currTkn.stype != Token::shortTtype::eof )
    {     
        switch (currTkn.stype)
        {
        case Token::shortTtype::num:
            chdr->setFormat(currTkn.posStartOfWord, currTkn.posEndOfWord, 
                            Format::charColor::green);
            break;
        
        case Token::shortTtype::keyword:
            chdr->setFormat(currTkn.posStartOfWord, currTkn.posEndOfWord,
                            Format::charColor::blue);
            break;
        
        case Token::shortTtype::id:
            chdr->setFormat(currTkn.posStartOfWord, currTkn.posEndOfWord,
                            Format::charColor::turquoise);
            break;
        
        default: chdr->setFormat(currTkn.posStartOfWord, currTkn.posEndOfWord,
                            Format::charColor::white);
        }
        currTkn = lex->scan();
    }
    return *chdr;
}

File: lab4/java_solution/highlightingtext/Highlighter.java
// Highlighter.java
package highlightingtext;
import java.util.List;

public class Highlighter {
    private Document chdr;
    private Lexer lex;

    public Highlighter(Document chs) {
        this.chdr = chs;
        this.lex = new Lexer(chs);
    }

    public Highlighter(Document chs, Lexer l) {
        this.chdr = chs;
        this.lex = l;
    }

    public void addKeywords(List<String> kws) {
        for (String kw : kws) {
            lex.addKeyword(kw);
        }
    }

    public Document highlightCharStream() {
        Token currTkn = lex.scan();
        while (currTkn.getStype() != Token.ShortTtype.EOF) {
            switch (currTkn.getStype()) {
                case NUM:
                    chdr.setFormat(currTkn.getPosStartOfWord(), currTkn.getPosEndOfWord(),
                            Format.CharColor.GREEN);
                    break;
                case KEYWORD:
                    chdr.setFormat(currTkn.getPosStartOfWord(), currTkn.getPosEndOfWord(),
                            Format.CharColor.BLUE);
                    break;
                case ID:
                    chdr.setFormat(currTkn.getPosStartOfWord(), currTkn.getPosEndOfWord(),
                            Format.CharColor.RED);
                    break;
                default:
                    chdr.setFormat(currTkn.getPosStartOfWord(), currTkn.getPosEndOfWord(),
                            Format.CharColor.WHITE);
            }
            currTkn = lex.scan();
        }
        return chdr;
    }
}